sweepbatcher/sweep_batch.go

```diff
@@ -561,46 +520,151 @@ func (b *batch) addSweep(ctx context.Context, sweep *sweep) (bool, error) {
+				s.outpoint, b.id, err)
+
+			return false, nil
+		}
+		if existing {
+			numExisting++
+		} else {
+			numNew++
+		}
+	}
+
+	// Make sure the whole group is either new or existing. If this is not
+	// the case, this might be a bug, so print a warning.
+	if numExisting > 0 && numNew > 0 {
+		b.Warnf("There are %d existing and %d new sweeps among the "+
+			"group. They must not be mixed.", numExisting, numNew)
+
+		return false, nil
+	}
+
+	// Make sure all the sweeps spend different outpoints.
+	outpointsSet := make(map[wire.OutPoint]struct{}, len(sweeps))
+	for _, s := range sweeps {
+		if _, has := outpointsSet[s.outpoint]; has {
+			b.Warnf("Multiple sweeps spend outpoint %v", s.outpoint)
 
 			return false, nil
 		}
+		outpointsSet[s.outpoint] = struct{}{}
 	}
 
 	// Past this point we know that a new incoming sweep passes the
 	// acceptance criteria and is now ready to be added to this batch.
 
-	// If this is the first sweep being added to the batch, make it the
-	// primary sweep.
-	if b.primarySweepID == zeroSweepID {
-		b.primarySweepID = sweep.outpoint
-		b.cfg.batchConfTarget = sweep.confTarget
-		b.rbfCache.FeeRate = sweep.minFeeRate
-		b.rbfCache.SkipNextBump = true
-
-		// We also need to start the spend monitor for this new primary
-		// sweep.
-		err := b.monitorSpend(ctx, *sweep)
-		if err != nil {
-			return false, err
+	// For an existing group, update the sweeps in the batch.
+	if numExisting == len(sweeps) {
+		for _, s := range sweeps {
+			oldSweep, ok := b.sweeps[s.outpoint]
+			if !ok {
+				return false, fmt.Errorf("sweep %v not found "+
+					"in batch %d", s.outpoint, b.id)
+			}
+
+			// Preserve coopFailed value not to forget about
+			// cooperative spending failure in this sweep.
+			tmp := *s
+			tmp.coopFailed = oldSweep.coopFailed
+
+			// If the sweep was resumed from storage, and the swap
+			// requested to sweep again, a new sweep notifier will
+			// be created by the swap. By re-assigning to the
+			// batch's sweep we make sure that everything, including
+			// the notifier, is up to date.
+			b.sweeps[s.outpoint] = tmp
+
+			// If this is the primary sweep, we also need to update
+			// the batch's confirmation target and fee rate.
+			if b.primarySweepID == s.outpoint {
+				b.cfg.batchConfTarget = s.confTarget
+				b.rbfCache.SkipNextBump = true
+			}
+
+			// Update batch's fee rate to be greater than or equal
+			// to minFeeRate of the sweep. Make sure batch's fee
+			// rate does not decrease (otherwise it won't pass RBF
+			// rules and won't be broadcasted) and that it is not
+			// lower that minFeeRate of other sweeps (so it is
+			// applied).
+			if b.rbfCache.FeeRate < s.minFeeRate {
+				b.rbfCache.FeeRate = s.minFeeRate
+			}
 		}
+
+		return true, nil
 	}
 
-	// Add the sweep to the batch's sweeps.
-	b.Infof("adding sweep %x", sweep.swapHash[:6])
-	b.sweeps[sweep.outpoint] = *sweep
+	// Here is the code to add new sweeps to a batch.
```

Comment:
should we check here that `len(sweeps)` `is numNew` and return an error otherwise.

Reply:
This logically follows from the way numExisting and numNew were calculated and from the fact that we have checked above that only one of these categories has any sweeps in it.

I added the following double-check just in case:

```go
        } else if numNew != len(sweeps) {
                // Sanity check: all the sweeps must be either existing or new.
                // We have checked this above, let's check here as well.
                return false, fmt.Errorf("bug in numExisting and numNew logic:"+
                        " numExisting=%d, numNew=%d, len(sweeps)=%d, "+
                        "len(b.sweeps)=%d", numExisting, numNew, len(sweeps),
                        len(b.sweeps))
```